File: src/safetensors.py
"""The loading dock: a .safetensors file on disk -> named bf16 tensors in memory, zero-copy.

File layout:  [8-byte little-endian header length][JSON catalog][raw tensor bytes, back-to-back]
"""
import json
import mmap
import struct

import torch

# safetensors dtype string -> torch dtype (only the ones we expect to meet)
DTYPES = {
    "BF16": torch.bfloat16,
    "F16": torch.float16,
    "F32": torch.float32,
}


class SafetensorsFile:
    def __init__(self, path: str):
        self._file = open(path, "rb")

        # Memory-map the file: the OS maps its bytes into our address space and
        # pulls pages from the SSD only when first touched. Nothing is read up front.
        # ACCESS_COPY = private copy-on-write mapping: readable, writable in our
        # process only, never written back to disk.
        self._mm = mmap.mmap(
            self._file.fileno(),
            0,
            access=mmap.ACCESS_COPY,
        )

        # First 8 bytes: the catalog's length. '<Q' = little-endian unsigned 64-bit int.
        header_len = struct.unpack("<Q", self._mm[:8])[0]

        # Next header_len bytes: the JSON catalog (name -> dtype, shape, data_offsets).
        header = json.loads(
            self._mm[8 : 8 + header_len]
        )

        # Keep every real tensor; drop the one non-tensor entry.
        self._catalog = {
            name: info
            for name, info in header.items()
            if name != "__metadata__"
        }

        # Where the raw tensor bytes begin; catalog offsets are relative to this point.
        self._data_start = 8 + header_len

    def tensor_names(self) -> list[str]:
        return list(self._catalog)

    def info(self, name: str) -> dict:
        if name not in self._catalog:
            raise KeyError(f"no tensor named {name!r}")

        return self._catalog[name]

    def get(self, name: str) -> torch.Tensor:
        """Return a tensor that VIEWS the file bytes directly. No copy, no dtype conversion."""
        t = self.info(name)

        dtype = DTYPES[t["dtype"]]

        start, end = t["data_offsets"]

        # bytes / bytes-per-element = element count (2 for bf16)
        numel = (
            (end - start)
            // torch.tensor([], dtype=dtype).element_size()
        )

        # Point a tensor at the mmap'd bytes: warehouse door + catalog offset = file position.
        flat = torch.frombuffer(
            self._mm,
            dtype=dtype,
            count=numel,
            offset=self._data_start + start,
        )

        return flat.view(*t["shape"])

    def integrity_check(self) -> bool:
        """The catalog must account for every byte: last end offset == end of file."""
        max_end = max(
            t["data_offsets"][1]
            for t in self._catalog.values()
        )

        return self._data_start + max_end == len(self._mm)

File: src/safetensors.ts
import { readFileSync } from "node:fs";

// ---- The contract: what one catalog entry looks like ----
export interface TensorInfo {
  dtype: string;
  shape: number[];
  data_offsets: [number, number]; // always exactly [start, end], relative to the data section
}

// ---- What the engine receives: a shape + the numbers ----
export interface Tensor {
  name: string;
  shape: number[];
  data: Float32Array;
}

// ---- bf16 -> fp32: bf16 is fp32's top 16 bits, so glue 16 zero bits underneath ----
function bf16ToF32(bytes: Buffer, start: number, count: number): Float32Array {
  const out = new Float32Array(count);
  const outBits = new Uint32Array(out.buffer); // second lens over the SAME memory
  for (let i = 0; i < count; i++) {
    const u16 = bytes.readUInt16LE(start + i * 2);
    outBits[i] = u16 << 16;
  }
  return out;
}

// ---- The loading dock: file on disk -> named tensors in memory ----
export class SafetensorsFile {
  private readonly buf: Buffer;
  private readonly catalog = new Map<string, TensorInfo>();
  private readonly dataStart: number;

  constructor(path: string) {
    this.buf = readFileSync(path);
    const headerLen = Number(this.buf.readBigUInt64LE(0));
    const header = JSON.parse(this.buf.subarray(8, 8 + headerLen).toString("utf8"));
    for (const [name, info] of Object.entries(header)) {
      if (name === "__metadata__") continue;
      this.catalog.set(name, info as TensorInfo);
    }
    this.dataStart = 8 + headerLen;
  }

  tensorNames(): string[] {
    return [...this.catalog.keys()];
  }

  info(name: string): TensorInfo {
    const t = this.catalog.get(name);
    if (!t) throw new Error(`no tensor named "${name}"`);
    return t;
  }

  getTensor(name: string): Tensor {
    const t = this.info(name);
    if (t.dtype !== "BF16") throw new Error(`${name}: unsupported dtype ${t.dtype}`);
    const [start, end] = t.data_offsets;
    const data = bf16ToF32(this.buf, this.dataStart + start, (end - start) / 2);
    return { name, shape: t.shape, data };
  }

  // Every byte of the data section must be claimed by exactly the catalog: last end offset == file end
  integrityCheck(): boolean {
    let maxEnd = 0;
    for (const t of this.catalog.values()) maxEnd = Math.max(maxEnd, t.data_offsets[1]);
    return this.dataStart + maxEnd === this.buf.length;
  }
}
